File: library/src/main/java/com/pengrad/telegrambot/request/SendInvoice.java
package com.pengrad.telegrambot.request;

import com.pengrad.telegrambot.model.request.LabeledPrice;

/**
 * Stas Parshin
 * 24 May 2017
 */
public class SendInvoice extends AbstractSendRequest<SendInvoice> {

    public SendInvoice(Long chatId, String title, String description, String payload, String providerToken,
                        String currency, LabeledPrice... prices) {
        super(chatId);
        add("title", title).add("description", description).add("payload", payload).add("provider_token", providerToken)
                .add("currency", currency).add("prices", prices);
    }

    // Backward compatibility: API 5.2, parameter "start_parameter" became optional
    public SendInvoice(Long chatId, String title, String description, String payload, String providerToken,
                       String startParameter, String currency, LabeledPrice... prices) {
        this(chatId, title, description, payload, providerToken, currency, prices);
        add("start_parameter", startParameter);
    }

    public SendInvoice providerData(String providerData) {
        return add("provider_data", providerData);
    }

    public SendInvoice photoUrl(String photoUrl) {
        return add("photo_url", photoUrl);
    }

    public SendInvoice photoSize(Integer photoSize) {
        return add("photo_size", photoSize);
    }

    public SendInvoice photoWidth(Integer photoWidth) {
        return add("photo_width", photoWidth);
    }

    public SendInvoice photoHeight(Integer photoHeight) {
        return add("photo_height", photoHeight);
    }

    public SendInvoice needName(boolean needName) {
        return add("need_name", needName);
    }

    public SendInvoice needPhoneNumber(boolean needPhoneNumber) {
        return add("need_phone_number", needPhoneNumber);
    }

    public SendInvoice needEmail(boolean needEmail) {
        return add("need_email", needEmail);
    }

    public SendInvoice needShippingAddress(boolean needShippingAddress) {
        return add("need_shipping_address", needShippingAddress);
    }

    public SendInvoice isFlexible(boolean isFlexible) {
        return add("is_flexible", isFlexible);
    }

    public SendInvoice startParameter(String startParameter) {
        return add("start_parameter", startParameter);
    }

    /**
     * 
     * @param maxTipAmount The maximum accepted amount for tips in the smallest units of the currency
     */
    public SendInvoice maxTipAmount(int maxTipAmount) {
        return add("max_tip_amount", maxTipAmount);
    }

    /**
     * 
     * @param suggestedTipAmounts An array of suggested amounts of tip in the smallest units of the currency. At most 4 suggested tip amounts can be specified. The suggested tip amounts must be positive, passed in a strictly increased order and must not exceed max_tip_amount.
     */
    public SendInvoice suggestedTipAmounts(Integer[] suggestedTipAmounts) {
        return add("suggested_tip_amounts", suggestedTipAmounts);
    }

    public SendInvoice sendEmailToProvider(boolean sendEmailToProvider) { return add("send_email_to_provider", sendEmailToProvider); }

    public SendInvoice sendPhoneNumberToProvider(boolean sendPhoneNumberToProvider) { return add("send_phone_number_to_provider", sendPhoneNumberToProvider); }

}

File: library/src/main/java/com/pengrad/telegrambot/request/SendInvoice.kt
package com.pengrad.telegrambot.request

import com.pengrad.telegrambot.model.request.LabeledPrice
import com.pengrad.telegrambot.utility.kotlin.checkDeprecatedConstructorParameters
import com.pengrad.telegrambot.utility.kotlin.optionalRequestParameter
import com.pengrad.telegrambot.utility.kotlin.requestParameter

class SendInvoice private constructor(
    chatId: Long? = null,
    channelUsername: String? = null,

    title: String,
    description: String,
    payload: String,
    currency: String,
    prices: List<LabeledPrice>
) : KAbstractSendRequest<SendInvoice>(
    chatId = chatId,
    channelUsername = channelUsername,
) {

    constructor(
        chatId: Long,
        title: String,
        description: String,
        payload: String,
        currency: String,
        prices: List<LabeledPrice>
    ) : this(
        chatId = chatId,
        channelUsername = null,
        title =  title,
        description = description,
        payload = payload,
        currency = currency,
        prices = prices
    )

    constructor(
        channelUsername: String,
        title: String,
        description: String,
        payload: String,
        currency: String,
        prices: List<LabeledPrice>
    ) : this(
        chatId = null,
        channelUsername = channelUsername,
        title =  title,
        description = description,
        payload = payload,
        currency = currency,
        prices = prices
    )

    @Deprecated("Use constructor with chatId or channelUsername instead", ReplaceWith("SendInvoice(chatId, title, description, payload, currency, prices.toList())"))
    constructor(
        chatId: Any,
        title: String,
        description: String,
        payload: String,
        currency: String,
        vararg prices: LabeledPrice
    ) : this(
        chatId = (chatId as? Number)?.toLong(),
        channelUsername = chatId as? String,
        title =  title,
        description = description,
        payload = payload,
        currency = currency,
        prices = prices.toList()
    ) {
        checkDeprecatedConstructorParameters()
    }

    val title: String by requestParameter(title)
    val description: String by requestParameter(description)
    val payload: String by requestParameter(payload)
    val currency: String by requestParameter(currency)
    val prices: List<LabeledPrice> by requestParameter(prices)

    var providerToken: String? by optionalRequestParameter()
    var providerData: String? by optionalRequestParameter()

    var maxTipAmount: Int? by optionalRequestParameter()
    var suggestedTipAmounts: List<Int>? by optionalRequestParameter()

    var photoUrl: String? by optionalRequestParameter()
    var photoSize: Int? by optionalRequestParameter()
    var photoWidth: Int? by optionalRequestParameter()
    var photoHeight: Int? by optionalRequestParameter()

    var needName: Boolean? by optionalRequestParameter()
    var needPhoneNumber: Boolean? by optionalRequestParameter()
    var needEmail: Boolean? by optionalRequestParameter()
    var needShippingAddress: Boolean? by optionalRequestParameter()

    var sendEmailToProvider: Boolean? by optionalRequestParameter()
    var sendPhoneNumberToProvider: Boolean? by optionalRequestParameter()

    var isFlexible: Boolean? by optionalRequestParameter()
    var subscriptionPeriod: Int? by optionalRequestParameter()
    var startParameter: String? by optionalRequestParameter()

    fun providerToken(providerToken: String) = applySelf { this.providerToken = providerToken }

    fun providerData(providerData: String) = applySelf { this.providerData = providerData }

    fun maxTipAmount(maxTipAmount: Int) = applySelf { this.maxTipAmount = maxTipAmount }

    fun suggestedTipAmounts(suggestedTipAmounts: List<Int>) = applySelf { this.suggestedTipAmounts = suggestedTipAmounts }

    fun suggestedTipAmounts(suggestedTipAmounts: Array<Int>) = suggestedTipAmounts(suggestedTipAmounts.toList())

    fun suggestedTipAmounts(vararg suggestedTipAmounts: Int) = suggestedTipAmounts(suggestedTipAmounts.toList())

    fun photoUrl(photoUrl: String) = applySelf { this.photoUrl = photoUrl }

    fun photoSize(photoSize: Int) = applySelf { this.photoSize = photoSize }

    fun photoWidth(photoWidth: Int) = applySelf { this.photoWidth = photoWidth }

    fun photoHeight(photoHeight: Int) = applySelf { this.photoHeight = photoHeight }

    fun needName(needName: Boolean) = applySelf { this.needName = needName }

    fun needPhoneNumber(needPhoneNumber: Boolean) = applySelf { this.needPhoneNumber = needPhoneNumber }

    fun needEmail(needEmail: Boolean) = applySelf { this.needEmail = needEmail }

    fun needShippingAddress(needShippingAddress: Boolean) = applySelf { this.needShippingAddress = needShippingAddress }

    fun sendEmailToProvider(sendEmailToProvider: Boolean) = applySelf { this.sendEmailToProvider = sendEmailToProvider }

    fun sendPhoneNumberToProvider(sendPhoneNumberToProvider: Boolean) = applySelf { this.sendPhoneNumberToProvider = sendPhoneNumberToProvider }

    fun isFlexible(isFlexible: Boolean) = applySelf { this.isFlexible = isFlexible }

    fun startParameter(startParameter: String) = applySelf { this.startParameter = startParameter }

}